secret_store/src/listener/service_contract_listener.rs

```diff
@@ -0,0 +1,699 @@
+use std::thread;
+use parking_lot::{Mutex, Condvar};	
+use ethcore::client::ChainNotify;
+use ethkey::{Random, Generator, Public, sign};
+use bytes::Bytes;
+use bigint::hash::H256;
+use bigint::prelude::U256;
+use key_server_set::KeyServerSet;
+use key_server_cluster::{ClusterClient, ClusterSessionsListener, ClusterSession};
+use key_server_cluster::generation_session::SessionImpl as GenerationSession;
+use key_storage::KeyStorage;
+use listener::service_contract::ServiceContract;
+use {ServerKeyId, NodeKeyPair, KeyServer};
+
+/// Retry interval (in blocks). Every RETRY_INTEVAL_BLOCKS blocks each KeyServer reads pending requests from
+/// service contract && tries to re-execute. The reason to have this mechanism is primarily because keys
+/// servers set change takes a lot of time + there could be some races, when blocks are coming to different
+/// KS at different times. This isn't intended to fix && respond to general session errors!
+const RETRY_INTEVAL_BLOCKS: usize = 30;
+
+/// Max failed retry requests (in single retry interval). The reason behind this constant is that if several
+/// pending requests have failed, then most probably other will fail too.
+const MAX_FAILED_RETRY_REQUESTS: usize = 1;
+
+/// SecretStore <-> Authority connector responsible for:
+/// 1. listening for new requests on SecretStore contract
+/// 2. redirecting requests to key server
+/// 3. publishing response on SecretStore contract
+pub struct ServiceContractListener {
+	/// Service contract listener data.
+	data: Arc<ServiceContractListenerData>,
+	/// Service thread handle.
+	service_handle: Option<thread::JoinHandle<()>>,
+}
+
+/// Service contract listener parameters.
+pub struct ServiceContractListenerParams {
+	/// Service contract.
+	pub contract: Arc<ServiceContract>,
+	/// Key server reference.
+	pub key_server: Arc<KeyServer>,
+	/// This node key pair.
+	pub self_key_pair: Arc<NodeKeyPair>,
+	/// Key servers set.
+	pub key_server_set: Arc<KeyServerSet>,
+	/// Cluster reference.
+	pub cluster: Arc<ClusterClient>,
+	/// Key storage reference.
+	pub key_storage: Arc<KeyStorage>,
+}
+
+/// Service contract listener data.
+struct ServiceContractListenerData {
+	/// Blocks since last retry.
+	pub last_retry: AtomicUsize,
+	/// Retry-related data.
+	pub retry_data: Mutex<ServiceContractRetryData>,
+	/// Service tasks queue.
+	pub tasks_queue: Arc<TasksQueue>,
+	/// Service contract.
+	pub contract: Arc<ServiceContract>,
+	/// Key server reference.
+	pub key_server: Arc<KeyServer>,
+	/// This node key pair.
+	pub self_key_pair: Arc<NodeKeyPair>,
+	/// Key servers set.
+	pub key_server_set: Arc<KeyServerSet>,
+	/// Key storage reference.
+	pub key_storage: Arc<KeyStorage>,
+
+}
+
+/// Retry-related data.
+#[derive(Default)]
+struct ServiceContractRetryData {
+	/// Server keys, which we have generated (or tried to generate) since last retry moment.
+	pub generated_keys: HashSet<ServerKeyId>,
+}
+
+/// Service tasks queue.
+struct TasksQueue {
```

Comment:
Why not `std::sync::mpsc::channel()` or `sync_channel(1024)` to prevent creating unbounded queue?

Reply:
Ok, I see that you are pushing multiple tasks at once, this structure is probably better since it won't be doing unecessary allocations.

Reply:
Actually the main reason behind this queue is an ability to shutdown processing thread right before queued tasks are processed. I mean - if there are 1000 queued tasks, then we could not push `Shutdown` task to the front of `mpsc::queue` && shutdown right after current task execution will be completed. Am I right that it is impossible with `mpsc::queue` (I could only think of dropping `transmitter` => this could lead to an error in next `receiver.wait()`, but this seems like a hack to me)?

Reply:
Dropping the transmitter is the only way, but afair `receiver.wait()` cannot fail in any other case, so I think this is the way it was intended to be used.

Reply:
Anyway - I've left the `TasksQueue` class, because it is useful in tests.